narwhal/consensus/src/bullshark.rs

```diff
@@ -1,27 +1,45 @@
 // Copyright (c) 2021, Facebook, Inc. and its affiliates
 // Copyright (c) Mysten Labs, Inc.
 // SPDX-License-Identifier: Apache-2.0
+use crate::metrics::ConsensusMetrics;
 use crate::{
     consensus::{ConsensusProtocol, ConsensusState, Dag},
     utils, ConsensusOutput,
 };
 use config::{Committee, Stake};
 use fastcrypto::{hash::Hash, traits::EncodeDecodeBase64};
 use std::{collections::BTreeSet, sync::Arc};
+use tokio::time::Instant;
 use tracing::{debug, error};
 use types::{Certificate, CertificateDigest, ConsensusStore, Round, SequenceNumber, StoreResult};
 
 #[cfg(test)]
 #[path = "tests/bullshark_tests.rs"]
 pub mod bullshark_tests;
 
+#[derive(Default)]
+struct LastRound {
+    _round: Round,
+    leader_found: bool,
+    leader_has_support: bool,
+}
+
 pub struct Bullshark {
     /// The committee information.
     pub committee: Committee,
     /// Persistent storage to safe ensure crash-recovery.
     pub store: Arc<ConsensusStore>,
     /// The depth of the garbage collector.
     pub gc_depth: Round,
+
+    pub metrics: Arc<ConsensusMetrics>,
+    /// The last time we had a successful leader election
+    /// which had enough support
+    last_leader_election_timestamp: Instant,
+    /// The last round for which a successful leader election took place
+    last_election_round: LastRound,
```

Comment:
nit - can you match the names whether you use "last_leader_election" or "last_election"? (whichever you choose)